util/gpool/spmc/spmspool.go

```diff
@@ -0,0 +1,386 @@
+}
+
+// AddProducer is to add producer.
+func (p *Pool[T, U, C, CT, TF]) AddProducer(producer func() (T, error), constArg C, contextFn TF, options ...TaskOption) (<-chan U, pooltask.TaskController[T, U, C, CT, TF]) {
+	opt := loadTaskOptions(options...)
+	taskID := p.NewTaskID()
+	var wg sync.WaitGroup
+	result := make(chan U, opt.ResultChanLen)
+	closeCh := make(chan struct{})
+	tc := pooltask.NewTaskController[T, U, C, CT, TF](p, taskID, closeCh, &wg)
+	taskCh := make(chan pooltask.Task[T], opt.TaskChanLen)
+	for i := 0; i < opt.Concurrency; i++ {
+		err := p.run()
+		if err == gpool.ErrPoolClosed {
+			break
+		}
+		taskBox := pooltask.NewTaskBox[T, U, C, CT, TF](constArg, contextFn, &wg, taskCh, result, taskID)
+		p.taskCh <- &taskBox
+	}
+	go func() {
+		defer func() {
+			close(closeCh)
+			close(taskCh)
+		}()
+		for {
+			task, err := producer()
+			if err != nil {
+				return
+			}
+			wg.Add(1)
+			if err != nil {
+				log.Fatal("fail to static", zap.Error(err))
+			}
+			t := pooltask.Task[T]{
+				Task: task,
+			}
+			taskCh <- t
+		}
+	}()
+	return result, tc
+}
+
+func (p *Pool[T, U, C, CT, TF]) run() error {
+	if p.IsClosed() {
+		return gpool.ErrPoolClosed
+	}
+	var w *goWorker[T, U, C, CT, TF]
+	if w = p.retrieveWorker(); w == nil {
+		return gpool.ErrPoolOverload
+	}
+	return nil
+}
+
+// retrieveWorker returns an available worker to run the tasks.
+func (p *Pool[T, U, C, CT, TF]) retrieveWorker() (w *goWorker[T, U, C, CT, TF]) {
+	spawnWorker := func() {
+		w = p.workerCache.Get().(*goWorker[T, U, C, CT, TF])
+		w.taskBoxCh = p.taskCh
+		w.run()
+	}
+
+	p.lock.Lock()
+
+	w = p.workers.detach()
+	if w != nil { // first try to fetch the worker from the queue
+		p.lock.Unlock()
+	} else if capacity := p.Cap(); capacity == -1 || capacity > p.Running() {
+		// if the worker queue is empty and we don't run out of the pool capacity,
+		// then just spawn a new worker goroutine.
+		p.lock.Unlock()
+		spawnWorker()
+	} else { // otherwise, we'll have to keep them blocked and wait for at least one worker to be put back into pool.
+		if p.options.Nonblocking {
+			p.lock.Unlock()
+			return
+		}
+	retry:
+		if p.options.MaxBlockingTasks != 0 && p.Waiting() >= p.options.MaxBlockingTasks {
+			p.lock.Unlock()
+			return
+		}
+		p.addWaiting(1)
+		p.cond.Wait() // block and wait for an available worker
+		p.addWaiting(-1)
+
+		if p.IsClosed() {
+			p.lock.Unlock()
+			return
+		}
+
+		var nw int
+		if nw = p.Running(); nw == 0 { // awakened by the scavenger
+			p.lock.Unlock()
+			spawnWorker()
+			return
+		}
+		if w = p.workers.detach(); w == nil {
+			if nw < p.Cap() {
+				p.lock.Unlock()
+				spawnWorker()
+				return
+			}
+			goto retry
+		}
+		p.lock.Unlock()
+	}
+	return
+}
+
+// revertWorker puts a worker back into free pool, recycling the goroutines.
+func (p *Pool[T, U, C, CT, TF]) revertWorker(worker *goWorker[T, U, C, CT, TF]) bool {
+	if capacity := p.Cap(); (capacity > 0 && p.Running() > capacity) || p.IsClosed() {
```

Comment:
`p.IsClosed()` is called in line371, so could we remove it here？

Reply:
In fact, It's a bug fix to avoid memory leaks.

Reply:
It is a data race problem between ```Release``` and ```revertWorker```. it is necessary.